### PR TITLE
Enable flake8-bugbear rule

### PR DESCRIPTION
**Pull request by CoolCat467**
This PR enables the flake8-bugbear rule and fixes all the existing issues that were raised.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -23,9 +23,11 @@ src = ["trio", "notes-to-self"]
 
 select = [
   "RUF",  # Ruff-specific rules
-  "E",  # Error
   "F",  # pyflakes
+  "E",  # Error
+  "W",  # Warning
   "I",  # isort
+  "B",  # flake8-bugbear
   "YTT",  # flake8-2020
```
Comment (jakkdl) [inline pyproject.toml:31]: you moved `"E"` ... probably want this list sorted?

File: pyproject.toml [hunk 2]
```diff
@@ -54,6 +56,9 @@ extend-exclude = [
 [tool.ruff.isort]
 combine-as-imports = true
 
+[tool.ruff.flake8-annotations]
+ignore-fully-untyped = true
+
```
Comment (jakkdl) [inline pyproject.toml:61]: this is unrelated to this PR

File: trio/_core/_io_windows.py [hunk 3]
```diff
@@ -368,7 +368,7 @@ def _get_base_socket(sock: _HasFileNo | int | Handle) -> Handle:
                     "https://github.com/python-trio/trio/issues/new, "
                     "and include the output of running: "
                     "netsh winsock show catalog"
-                )
+                ) from ex
```
Comment (jakkdl) [inline trio/_core/_io_windows.py:371]: this is a functionality change, though I have no clue if it's bad.

File: trio/_abc.py [hunk 4]
```diff
@@ -80,9 +80,11 @@ class Instrument(metaclass=ABCMeta):
 
     def before_run(self) -> None:
         """Called at the beginning of :func:`trio.run`."""
+        return
```
Comment (jakkdl) [inline trio/_abc.py:83]: lots of codecov misses, maybe `# pragma: no cover` on these methods?

File: trio/_dtls.py [hunk 5]
```diff
@@ -841,10 +841,7 @@ def __init__(self, endpoint: DTLSEndpoint, peer_address: Any, ctx: Context):
         # support and isn't useful anyway -- especially for DTLS where it's equivalent
         # to just performing a new handshake.
         ctx.set_options(
-            (
-                SSL.OP_NO_QUERY_MTU
-                | SSL.OP_NO_RENEGOTIATION  # type: ignore[attr-defined]
-            )
+            SSL.OP_NO_QUERY_MTU | SSL.OP_NO_RENEGOTIATION  # type: ignore[attr-defined]
```
Comment (jakkdl) [inline trio/_dtls.py:844]: is this formatting change due to a version bump?

File: trio/_tests/test_path.py [hunk 6]
```diff
@@ -114,7 +114,7 @@ async def test_compare_async_stat_methods(method_name):
 
 async def test_invalid_name_not_wrapped(path):
     with pytest.raises(AttributeError):
-        getattr(path, "invalid_fake_attr")
+        path.invalid_fake_attr  # noqa: B018
```
Comment (jakkdl) [inline trio/_tests/test_path.py:117]: hmmm. Does this not raise a type error that would also need to be silenced?
(Also probably good with a comment that B018 is `useless expression`, as mentioned in other comments)

File: trio/_tests/test_socket.py [hunk 7]
```diff
@@ -1108,7 +1111,8 @@ async def receiver() -> None:
 async def test_many_sockets() -> None:
     total = 5000  # Must be more than MAX_AFD_GROUP_SIZE
     sockets = []
-    for x in range(total // 2):
+    x = 0
+    for x in range(total // 2):  # noqa: B007  # Used for check below
```
Comment (jakkdl) [inline trio/_tests/test_socket.py:1115]: > B007: Loop control variable not used within the loop body. If this is intended, start the name with an underscore.

can rename it instead of silencing the warning. Though I'm used to underscore-variables meaning they're unused so I'm not loving the convention, but /shrug

File: trio/_tests/test_threads.py [hunk 8]
```diff
@@ -463,7 +463,7 @@ async def run_thread(event):
         async with _core.open_nursery() as nursery:
             print("spawning")
             events = []
-            for i in range(COUNT):
+            for _i in range(COUNT):
```
Comment (jakkdl) [inline trio/_tests/test_threads.py:466]: to differentiate from B007-induced underscore I think you should use the variable name `_`, I think there were a few other instances earlier as well.

File: trio/_tests/test_tracing.py [hunk 9]
```diff
@@ -21,7 +21,7 @@ async def coro2_async_gen(event):
 
 
 async def coro3_async_gen(event: trio.Event) -> None:
-    async for x in coro2_async_gen(event):
+    async for _x in coro2_async_gen(event):
```
Comment (jakkdl) [inline trio/_tests/test_tracing.py:24]: `_`

File: trio/socket.py [hunk 10]
```diff
@@ -83,7 +83,7 @@
     IP_BIND_ADDRESS_NO_PORT: int
 else:
     try:
-        IP_BIND_ADDRESS_NO_PORT
+        IP_BIND_ADDRESS_NO_PORT  # noqa: B018  # "useless expression"
```
Comment (jakkdl) [inline trio/socket.py:86]: Oh yeah, can probably just do a search+replace for `# noqa: B018\n` -> `# noqa: B018 # "useless expression"`

**Review by jakkdl — APPROVED** [review]
Big approve on enabling bugbear, bunch of small style comments.

File: pyproject.toml (hunk 1 above)
Comment (jakkdl) [reply pyproject.toml:31]: We should consider running a toml-formatter in pre-commit that does it automatically

File: trio/_tests/test_path.py (hunk 6 above)
Comment (TeamSpen210) [reply trio/_tests/test_path.py:117]: No, it should just be an AttributeError from `__getattr__()`.

File: trio/_abc.py (hunk 4 above)
Comment (TeamSpen210) [reply trio/_abc.py:83]: Or actually, maybe just add a test that calls all of these? The fact that inheriting from `Instrument` gives you do-nothing implementations of all the hook methods (so they won't error out) is something we probably want to guarantee.

File: pyproject.toml (hunk 2 above)
Comment (CoolCat467) [reply pyproject.toml]: Removed in 97bcb65

File: trio/_abc.py (hunk 4 above)
Comment (CoolCat467) [reply trio/_abc.py:83]: How would you guarantee that a function does nothing?

File: trio/_dtls.py (hunk 5 above)
Comment (CoolCat467) [reply trio/_dtls.py:844]: It's changed by black funnily enough, not ruff.
Tested with reverting changes and running pre-commit locally.

File: trio/_abc.py (hunk 4 above)
Comment (TeamSpen210) [reply trio/_abc.py:83]: We just want to test that calling it works - it exists and doesn't raise an exception.

File: trio/socket.py (hunk 10 above)
Comment (CoolCat467) [reply trio/socket.py:86]: Changed in 4213618

File: trio/_tests/test_tracing.py (hunk 9 above)
Comment (CoolCat467) [reply trio/_tests/test_tracing.py]: Changed in 1147980

File: trio/_tests/test_threads.py (hunk 8 above)
Comment (CoolCat467) [reply trio/_tests/test_threads.py]: Changed in https://github.com/python-trio/trio/commit/11479806a721b0e27e6d4099eab91f59db247976

File: trio/_tests/test_socket.py (hunk 7 above)
Comment (CoolCat467) [reply trio/_tests/test_socket.py]: Changed in https://github.com/python-trio/trio/commit/11479806a721b0e27e6d4099eab91f59db247976

File: trio/_abc.py (hunk 4 above)
Comment (CoolCat467) [reply trio/_abc.py:83]: Added test in 9043b48

File: trio/_tests/test_path.py (hunk 6 above)
Comment (CoolCat467) [reply trio/_tests/test_path.py]: Changed to use `getattr` again in 2bd476d

File: trio/_core/_io_windows.py (hunk 3 above)
Comment (CoolCat467) [reply trio/_core/_io_windows.py:371]: My logic changing this was that maybe seeing the full OS error in the traceback might be helpful, but up for debate.

File: trio/_tests/test_abc.py [hunk 11]
```diff
@@ -4,6 +4,32 @@
...
+    mayonnaise = tabc.Instrument()
+
+    for method_name, args in attrs.items():
+        assert hasattr(mayonnaise, method_name)
+        method = getattr(mayonnaise, method_name)
+        assert callable(method)
+        try:
+            method(*args)
+        except Exception as exc:
+            raise AssertionError("Raised exception") from exc
```
Comment (TeamSpen210) [inline trio/_tests/test_abc.py:32]: No need to catch and re-raise, any exception is going to cause the test to fail.

Comment (CoolCat467) [reply trio/_tests/test_abc.py]: Changed in f9c18dc

File: trio/_core/_io_windows.py (hunk 3 above)
Comment (jakkdl) [reply trio/_core/_io_windows.py:371]: Can probably keep it in, but add a note in newsfragments?

File: trio/_tests/test_deprecate.py [hunk 12]
```diff
@@ -259,7 +259,7 @@ def test_module_with_deprecations(recwarn_always: pytest.WarningsRecorder) -> No
     assert "instead-string instead" in got.message.args[0]
 
     with pytest.raises(AttributeError):
-        module_with_deprecations.asdf  # type: ignore[attr-defined]  # noqa
+        module_with_deprecations.asdf  # type: ignore[attr-defined]  # noqa: B018
```
Comment (jakkdl) [inline trio/_tests/test_deprecate.py:262]: `# "useless expression"`

(I really dislike that flake8 doesn't have name-based noqas)

File: trio/testing/_fake_net.py [hunk 13]
```diff
@@ -33,7 +33,7 @@ def _family_for(ip: IPAddress) -> int:
         return trio.socket.AF_INET
     elif isinstance(ip, ipaddress.IPv6Address):
         return trio.socket.AF_INET6
-    assert False  # pragma: no cover
+    raise AssertionError()  # pragma: no cover
```
Comment (jakkdl) [inline trio/testing/_fake_net.py:36]: here

File: trio/testing/_fake_net.py [hunk 14]
```diff
@@ -42,7 +42,7 @@ def _wildcard_ip_for(family: int) -> IPAddress:
     elif family == trio.socket.AF_INET6:
         return ipaddress.ip_address("::")
     else:
-        assert False
+        raise AssertionError()
```
Comment (jakkdl) [inline trio/testing/_fake_net.py:45]: here

File: trio/testing/_fake_net.py [hunk 15]
```diff
@@ -51,7 +51,7 @@ def _localhost_ip_for(family: int) -> IPAddress:
     elif family == trio.socket.AF_INET6:
         return ipaddress.ip_address("::1")
     else:
-        assert False
+        raise AssertionError()
```
Comment (jakkdl) [inline trio/testing/_fake_net.py:54]: Add an error message to the three `AssertionError`s in this file - something like "unknown IP Address family". Or a `NotImplementedError` instead.
Also `# pragma: no cover`.

File: trio/_dtls.py [hunk 16]
```diff
@@ -1305,7 +1305,9 @@ async def handler(dtls_channel):
         try:
             self.socket.getsockname()
         except OSError:
-            raise RuntimeError("DTLS socket must be bound before it can serve")
+            raise RuntimeError(
+                "DTLS socket must be bound before it can serve"
+            ) from None
```
Comment (jakkdl) [inline trio/_dtls.py:1310]: If you're up for writing a test that triggers this it'd be lovely, otherwise just mark it as `# pragma: no cover` and a comment that a test should be written.

File: trio/_dtls.py (hunk 5 above)
Comment (jakkdl) [reply trio/_dtls.py:844]: Well, it's a good change - but I'm confused why it's in this PR and not in a separate PR that does the actual version bump in the committed config file.

File: trio/_core/_tests/test_multierror.py [hunk 17]
```diff
@@ -484,9 +486,9 @@ def test_pickle_multierror(protocol: int) -> None:
     my_except = ZeroDivisionError()
 
     try:
-        1 / 0
-    except ZeroDivisionError as e:
-        my_except = e
+        1 / 0  # noqa: B018  # Need exception to fire
+    except ZeroDivisionError as exc:
+        my_except = exc
```
Comment (jakkdl) [inline trio/_core/_tests/test_multierror.py:491]: *very* minor nit, but I think it's quite obvious why the statement is there, and what needs a comment is why the `noqa` is needed - which is once again `"useless statement"` - to not make anybody think that what's being complained about is the division by zero.

File: trio/_core/_run.py [hunk 18]
```diff
@@ -2745,7 +2746,7 @@ async def checkpoint_if_cancelled() -> None:
         task is task._runner.main_task and task._runner.ki_pending
     ):
         await _core.checkpoint()
-        assert False  # pragma: no cover
+        raise AssertionError()  # pragma: no cover
```
Comment (jakkdl) [inline trio/_core/_run.py:2749]: very minor nit, but would be great with a message - any message, in the exception. A standard one I use when I don't come up with anything better is "this should never happen".

File: trio/_core/_tests/test_guest_mode.py [hunk 19]
```diff
@@ -72,7 +72,7 @@ def done_callback(outcome):
             elif op == "unwrap":
                 return obj.unwrap()
             else:  # pragma: no cover
-                assert False
+                raise AssertionError()
```
Comment (jakkdl) [inline trio/_core/_tests/test_guest_mode.py:75]: same

File: trio/_core/_tests/test_guest_mode.py [hunk 20]
```diff
@@ -301,7 +301,7 @@ async def sit_in_wait_all_tasks_blocked(watb_cscope):
                 # wait_all_tasks_blocked should *not* return normally, but
                 # only by cancellation.
                 await trio.testing.wait_all_tasks_blocked(cushion=9999)
-                assert False  # pragma: no cover
+                raise AssertionError()  # pragma: no cover
```
Comment (jakkdl) [inline trio/_core/_tests/test_guest_mode.py:304]: same - though here we have a comment we could copy text from.

File: trio/_core/_tests/test_instrumentation.py [hunk 21]
```diff
@@ -236,7 +236,7 @@ async def main():
 def test_instrument_that_raises_on_getattr():
     class EvilInstrument:
         def task_exited(self, task):
-            assert False  # pragma: no cover
+            raise AssertionError()  # pragma: no cover
```
Comment (jakkdl) [inline trio/_core/_tests/test_instrumentation.py:239]: same

**Review by jakkdl — COMMENTED** [review]
It feels a bit shitty to put the burden of cleaning up old smelly code in an entirely separate PR that just happens to touch those lines, so feel free to ignore ones you don't want to bother with.

File: trio/_core/_tests/test_multierror.py (hunk 17 above)
Comment (TeamSpen210) [reply trio/_core/_tests/test_multierror.py:491]: Maybe just replace it with `raise ZeroDivisionError`?

File: trio/_core/_tests/test_instrumentation.py (hunk 21 above)
Comment (CoolCat467) [reply trio/_core/_tests/test_instrumentation.py]: Added reason text in a5afed7

File: trio/_core/_tests/test_guest_mode.py (hunk 20 above)
Comment (CoolCat467) [reply trio/_core/_tests/test_guest_mode.py]: Added reason text in a5afed7

File: trio/_core/_run.py (hunk 18 above)
Comment (CoolCat467) [reply trio/_core/_run.py]: Added reason text in a5afed7

File: trio/_core/_tests/test_guest_mode.py (hunk 19 above)
Comment (CoolCat467) [reply trio/_core/_tests/test_guest_mode.py]: Changed to `NotImplementedError` with reason in a5afed7

File: trio/_core/_tests/test_multierror.py (hunk 17 above)
Comment (CoolCat467) [reply trio/_core/_tests/test_multierror.py]: Changed noqa additional comment details in a5afed7

File: trio/_dtls.py (hunk 16 above)
Comment (CoolCat467) [reply trio/_dtls.py]: Added nocover comment and TODO comment in a5afed7

File: trio/testing/_fake_net.py (hunk 15 above)
Comment (CoolCat467) [reply trio/testing/_fake_net.py]: Changed in a5afed7 and added reason in e1fdebc8

File: trio/testing/_fake_net.py (hunk 14 above)
Comment (CoolCat467) [reply trio/testing/_fake_net.py]: Changed in a5afed7 and added reason in e1fdebc8

File: trio/testing/_fake_net.py (hunk 13 above)
Comment (CoolCat467) [reply trio/testing/_fake_net.py]: Changed in a5afed7 and added reason in e1fdebc8

File: trio/_tests/test_deprecate.py (hunk 12 above)
Comment (CoolCat467) [reply trio/_tests/test_deprecate.py]: Added to comment in a5afed7

File: trio/_core/_io_windows.py (hunk 3 above)
Comment (CoolCat467) [reply trio/_core/_io_windows.py:371]: Added not in f80b22a5

**Review by jakkdl — APPROVED** [review]
Looks great~
